pruning/manager_test.go

```diff
@@ -0,0 +1,274 @@
+		}
+
+		if manager.ShouldPruneAtHeight(curHeight) {
+			manager.ResetPruningHeights()
+			heightsToPruneMirror = make([]int64, 0)
+		}
+
+		// N.B.: There is no reason behind the choice of 3.
+		if curHeight%3 == 0 {
+			require.Equal(t, heightsToPruneMirror, manager.GetPruningHeights(), curHeightStr)
+			batch := db.NewBatch()
+			manager.FlushPruningHeights(batch)
+			require.NoError(t, batch.Write())
+			require.NoError(t, batch.Close())
+
+			manager.ResetPruningHeights()
+			require.Equal(t, make([]int64, 0), manager.GetPruningHeights(), curHeightStr)
+
+			err := manager.LoadPruningHeights(db)
+			require.NoError(t, err)
+			require.Equal(t, heightsToPruneMirror, manager.GetPruningHeights(), curHeightStr)
+		}
+	}
+}
+
+func Test_WithSnapshot(t *testing.T) {
+	manager := pruning.NewManager(log.NewNopLogger())
+	require.NotNil(t, manager)
+
+	curStrategy := types.NewCustomPruningOptions(10, 10)
+	
+	snapshotInterval := uint64(15)
+	manager.SetSnapshotInterval(snapshotInterval)
+
+	manager.SetOptions(curStrategy)
+	require.Equal(t, curStrategy, manager.GetOptions())
+
+	keepRecent := curStrategy.KeepRecent
+
+	heightsToPruneMirror := make([]int64, 0)
+
+	mx := sync.Mutex{}
+	snapshotHeightsToPruneMirror := list.New()
+
+	wg := sync.WaitGroup{}
+
+	for curHeight := int64(1); curHeight < 100000; curHeight++ {
+		mx.Lock()
+		handleHeightActual := manager.HandleHeight(curHeight)
+
+		curHeightStr := fmt.Sprintf("height: %d", curHeight)
+
+		if curHeight > int64(keepRecent) && (curHeight-int64(keepRecent))%int64(snapshotInterval) != 0 {
+			expectedHandleHeight := curHeight - int64(keepRecent)
+			require.Equal(t, expectedHandleHeight, handleHeightActual, curHeightStr)
+			heightsToPruneMirror = append(heightsToPruneMirror, expectedHandleHeight)
+		} else {
+			require.Equal(t, int64(0), handleHeightActual, curHeightStr)
+		}
+
+		actualHeightsToPrune := manager.GetPruningHeights()
+
+		var next *list.Element
+		for e := snapshotHeightsToPruneMirror.Front(); e != nil; e = next {
+			snapshotHeight := e.Value.(int64)
+			if snapshotHeight < curHeight-int64(keepRecent) {
+				heightsToPruneMirror = append(heightsToPruneMirror, snapshotHeight)
+
+				// We must get next before removing to be able to continue iterating.
+				next = e.Next()
+				snapshotHeightsToPruneMirror.Remove(e)
+			} else {
+				next = e.Next()
+			}
+		}
+
+		require.Equal(t, heightsToPruneMirror, actualHeightsToPrune, curHeightStr)
+		mx.Unlock()
+
+		if manager.ShouldPruneAtHeight(curHeight) {
+			manager.ResetPruningHeights()
+			heightsToPruneMirror = make([]int64, 0)
+		}
+
+		// Mimic taking snapshots in the background
+		if curHeight%int64(snapshotInterval) == 0 {
+			wg.Add(1)
+			go func(curHeightCp int64) {
+				time.Sleep(time.Nanosecond * 500)
+
+				mx.Lock()
+				manager.HandleHeightSnapshot(curHeightCp)
+				snapshotHeightsToPruneMirror.PushBack(curHeightCp)
+				mx.Unlock()
+				wg.Done()
+			}(curHeight)
+		}
+	}
+
+	wg.Wait()
```

Comment:
Please move this to the very top below where you declared wg and make it a defer wg.Wait()